src/librustc/infer/error_reporting/anon_anon_conflict.rs

```diff
@@ -0,0 +1,216 @@
+                    }
+                }
+            }
+            _ => {}
+        }
+        // walk the embedded contents: e.g., if we are visiting `Vec<&Foo>`,
+        // go on to visit `&Foo`
+        intravisit::walk_ty(self, arg);
+    }
+}
+
+impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
+    // This function calls the `visit_ty` method for the parameters
+    // corresponding to the anonymous regions. The `nested_visitor.found_type`
+    // contains the anonymous type.
+    pub fn find_anon_type(&self, region: Region<'tcx>, br: &ty::BoundRegion) -> Option<&hir::Ty> {
+        if self.is_suitable_anonymous_region(region).is_some() {
+            let def_id = self.is_suitable_anonymous_region(region).unwrap();
+            let node_id = self.tcx.hir.as_local_node_id(def_id).unwrap();
+            let ret_ty = self.tcx.type_of(def_id);
+            match ret_ty.sty {
+                ty::TyFnDef(_, _) => {
+                    match self.tcx.hir.get(node_id) {
+                        hir_map::NodeItem(it) => {
+                            match it.node {
+                                hir::ItemFn(ref fndecl, _, _, _, _, _) => {
+                                    fndecl
+                                        .inputs
+                                        .iter()
+                                        .filter_map(|arg| {
+                                            let mut nested_visitor = FindNestedTypeVisitor {
+                                                infcx: &self,
+                                                hir_map: &self.tcx.hir,
+                                                bound_region: *br,
+                                                found_type: None,
+                                            };
+                                            nested_visitor.visit_ty(&**arg);
+                                            if nested_visitor.found_type.is_some() {
+                                                return nested_visitor.found_type;
+                                            } else {
+                                                return None;
+                                            }
+                                        })
+                                        .next()
+                                }
+                                _ => None,
+                            }
+                        }
+                        _ => None,
+                    }
+                }
+                _ => None,
+            }
+        } else {
+            None
+        }
+    }
+
+    pub fn try_report_anon_anon_conflict(&self, error: &RegionResolutionError<'tcx>) -> bool {
+
+        let (span, sub, sup) = match *error {
+            ConcreteFailure(ref origin, sub, sup) => (origin.span(), sub, sup),
+            _ => return false, // inapplicable
+        };
+
+        // Determine whether the sub and sup consist of both anonymous (elided) regions.
```

Comment:
This could be written a bit more succinctly:

```rust
// Determine whether the sub and sup consist of both anonymous (elided) regions.
if let (Some(br1), Some(br2)) = (self.is_anonymous_region(sup),
                                 self.is_anonymous_region(sub)) {
    if let (Some(ty1), Some(t2)) = (self.find_anon_type(sup, &br1),
                                    self.find_anon_type(sub, &br2)) {
        if let (Some((anon_arg1, _, _, _)), Some((anon_arg2, _, _, _))) =
            (self.find_arg_with_anonymous_region(sup, sup),
             self.find_arg_with_anonymous_region(sub, sub))
        {
            let mut err = struct_span_err!(self.tcx.sess, span, E0623, "lifetime mismatch")
                .span_label(ty1.span,
                            format!("these references must have the same lifetime"))
                .span_label(ty2.span, format!(""))
            let span_label = if let (Some(var1), Some(var2)) = (anon_arg1.pat.simple_name(),
                                                                anon_arg2.pat.simple_name()) {
                err.span_label(span, format!("data from `{}` flows into `{}` here", var1, var2));
            }
            err.emit();
            return true;
         }
    }
    false
}
```